### ISSUE TITLE
[BUG] cudf.get_dummies on a large cardinality column is slow 

### ISSUE THREAD

**Issue by VibhuJawa** [rapidsai/cudf#8870]
**Describe the bug**
cudf.get_dummies on a large cardinality column is slow 

**Steps/Code to reproduce bug**:

CUDF
```python
n_rows = 1_200_000
big_cat_series_cudf = cudf.Series(cp.random.randint(low=0, high=5100, size=n_rows)).astype('category')
cudf_df = cudf.DataFrame({"big_cat_series":big_cat_series_cudf})
```

```
%%time 
one_hot_df_cudf = cudf.get_dummies(cudf_df,columns = ['big_cat_series'])
```
```
CPU times: user 21.6 s, sys: 27 s, total: 48.6 s
Wall time: 49.2 s
```
Pandas:

```python
n_rows = 1_200_000
big_cat_series_pd = pd.Series(np.random.randint(low=0, high=5100, size=n_rows)).astype('category')
pd_df = pd.DataFrame({"big_cat_series":big_cat_series_pd})
```
```python

%%time 
one_hot_df_pd = pd.get_dummies(pd_df,columns = ['big_cat_series'])
```

```

CPU times: user 56.7 s, sys: 3.37 s, total: 1min
Wall time: 1min

```

**Expected behavior**
I would expect it to be faster. 
**Environment overview (please complete the following information)**
 - Method of cuDF install: [conda]

```
cudf                      21.08.00a210727 cuda_11.2_py38_ga69a8a43b5_324    rapidsai-nightly
```

**Comment by isVoid** [comment]
This is kind of like a `wide table` problem. When supplied with cardinality 5100, the resulting dataframe needs 5100 columns. In generating these columns, each of the column requires two kernel launches (`==` and `fillna`) and results in a large amount of small kernal launches.

cuml seems to have a [OneHotEncoder](https://docs.rapids.ai/api/cuml/stable/api.html?highlight=one%20hot#cuml.preprocessing.OneHotEncoder), though I currently don't have a working cuml build to test it. Will test its efficiency once available.

**Comment by jrhemstad** [comment]
> This is kind of like a `wide table` problem. When supplied with cardinality 5100, the resulting dataframe needs 5100 columns. In generating these columns, each of the column requires two kernel launches (`==` and `fillna`) and results in a large amount of small kernal launches.

We could solve the many kernel launch issue by having a custom kernel in libcudf to do the encoding. However, my understanding is that the real issue with the "wide table" problem is the Python overhead in serially constructing the 5100 Python objects. We ran into this in the olden days with `transpose`. See https://github.com/rapidsai/cudf/pull/356 and https://github.com/rapidsai/cudf/issues/745

**Comment by isVoid** [comment]
Agreed. Python overhead is the dominating part in time consumption. Also analogous to groupby apply issue we recently faced.

**Comment by isVoid** [comment]
I'd be curious what's the need for `dtype` parameter for `Series.one_hot_encoding`? What are the cases that you may need anything other than `1/0` in the output? I'm asking because the post processing of 5100 columns into the expected dtype is quite non-trivial (for both python and c++).
